zjit/src/hir.rs

```diff
@@ -3571,23 +3571,235 @@ impl Function {
+            .insert("type", ty)
+            .build()
+    }
+
+    /// Helper function to make an Iongraph JSON "block".
+    /// `attributes` is specific to Spidermonkey, so it is left empty.
+    /// todo(aidenfoxivey): Calculate loopDepth
+    /// todo(aidenfoxivey): Calculate attributes
+    fn make_iongraph_block(ptr: u64, id: u64, predecessors: Vec<u64>, successors: Vec<u64>, instructions: Vec<Json>) -> Json {
+        Json::object()
+            .insert("ptr", ptr)
+            .insert("id", id)
+            .insert("loopDepth", 0)
+            .insert("attributes", Json::empty_array())
+            .insert("predecessors", Json::array(predecessors))
+            .insert("successors", Json::array(successors))
+            .insert("instructions", Json::array(instructions))
+            .build()
+    }
+
+    /// Helper function to make an Iongraph JSON "function".
+    /// Note that `lir` is unpopulated right now as ZJIT doesn't use its functionality.
+    fn make_iongraph_function(pass_name: &str, hir_blocks: Vec<Json>) -> Json {
+        Json::object()
+            .insert("name", pass_name)
+            .insert("mir", Json::object()
+                .insert("blocks", Json::array(hir_blocks))
+                .build()
+            )
+            .insert("lir", Json::object()
+                .insert("blocks", Json::empty_array())
+                .build()
+            )
+            .build()
+    }
+
+    /// Generate an iongraph JSON pass representation for this function.
+    pub fn to_iongraph_pass(&self, pass_name: &str) -> Json {
+        let mut ptr_map = PtrPrintMap::identity();
+        if cfg!(test) {
+            ptr_map.map_ptrs = true;
+        }
+
+        let mut hir_blocks = Vec::new();
+
+        // Push each block from the iteration in reverse post order to `hir_blocks`.
+        for block_id in self.rpo() {
+            // Create the block with instructions.
+            let block = &self.blocks[block_id.0];
+            let block_ptr = ptr_map.map_id(block_id.0 as u64);
+            let (mut block_successors, mut block_predecessors) = self.compute_successors_and_predecessors();
+            let predecessors = block_predecessors.remove(&block_id).unwrap_or_default();
+            let successors = block_successors.remove(&block_id).unwrap_or_default();
+            let mut instructions = Vec::new();
+
+            // Get parameter instructions.
+            for (idx, param_id) in block.params.iter().enumerate() {
+                let param_id = self.union_find.borrow().find_const(*param_id);
+                let insn_type = self.type_of(param_id);
+
+                let type_str = if insn_type.is_subtype(types::Empty) {
+                    String::new()
+                } else {
+                    format!("{}", insn_type.print(&ptr_map))
+                };
+
+                instructions.push(
+                    Self::make_iongraph_instr(
+                        ptr_map.map_id(idx as u64) as u64,
```

Comment:
Don't use the idx as a ptr; it's not unique. use `param_id`

Reply:
You also don't need to use `ptr_map` since the ID should be relatively stable (it's not a real ptr)